framework/audio/internal/midiplayer.cpp

```diff
@@ -239,131 +254,81 @@ bool Sequencer::sendEvents(tick_t fromTick, tick_t toTick)
         }
 
         ChanState& chState = m_chanStates[event.channel()];
-        if (chState.muted || SKIP_EVENTS.find(event.type()) != SKIP_EVENTS.end()) {
-            // noop
-        } else {
+        if (event && !chState.muted) {
             auto s = synth(event.channel());
             s->handleEvent(event);
             s->setIsActive(true);
+
+            if (event.isChannelVoice() && event.opcode() == midi::Event::Opcode::NoteOn) {
+                auto noteOff = event;
+                noteOff.setOpcode(midi::Event::Opcode::NoteOff);
+                m_noteCache[event.note()] = noteOff;
+            } else if (event.isChannelVoice() && event.opcode() == midi::Event::Opcode::NoteOff) {
+                m_noteCache[event.note()] = Event::NOOP();
+            }
         }
 
         ++pos;
     }
 
+    midiPortDataSender()->sendEvents(fromTick, toTick);

```

Comment:
This does not take into account the latency in the fifo buffer, i.e. it is ahead of what is actually being sound.
If we need to increase the buffer size, then this will become a serious problem.